accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+from two superinterfaces then the subinterface must resolve the
+conflict. The so-called diamond inheritance pattern can create the
+case where two superinterfaces have an _m_, but they are both declared by
+the same declaration (so `V` is a subinterface of `V1` and `V2`, and both
+`V1` and `V2` are subinterfaces of `V3`, and only `V3` declares _m_,
+in which case there is no conflict in `V`).*
+
+*Assume that _DV_ is an inline class declaration named `Inline`, and
+the inline type `V1`, declared by _DV1_, is a superinterface of
+_DV_. Let `m` be the name of a member of `V1`. If _DV_ also declares a
+member named `m` then the latter may be considered similar to a
+declaration that "overrides" the former.  However, it should be noted
+that inline method invocation is resolved statically, and hence there
+is no override relationship among the two in the traditional
+object-oriented sense (that is, it will never occur that the
+statically known declaration is the member of `V1`, and the member
+invoked at run time is the one in _DV_). A receiver with static
+type `V1` will invoke the declaration in _DV1_, and a receiver with
+static type `Inline` (or `Inline<...>`) will invoke the one in _DV_.*
+
+Hence, we use a different word to describe the relationship between a
+member named _m_ of a superinterface, and a member named _m_ which is
+declared by the subinterface: We say that the latter _redeclares_ the
+former.
+
+*In particular, if two different declarations of _m_ is inherited
+from two superinterface then the subinterface can resolve the conflict
+by redeclaring _m_.*
+
+*Note that there is no notion of having a 'correct override relation'
+here. With inline classes, any member signature can redeclare any
+other member signature with the same name, including the case where a
+method is redeclared by a getter, or vice versa. The reason for this
+is that no call site will resolve to one of several declarations at
+run time. Each invocation will statically resolve to one particular
+declaration, and this makes it possible to ensure that the invocation
+is type correct.*
+
+Assume that _DV_ is an inline class declaration, and that the inline
+types `V1` and `V2` are superinterfaces of _DV_. Let `M1` be the
+members of `V1`, and `M2` the members of `V2`. A compile-time error
+occurs if there is a member name `m` such that `V1` as well as `V2`
+has a member named `m`, and they are distinct declarations, and _DV_
+does not declare a member named `m`.  *In other words, a name clash
+among distinct "inherited" members is an error, but it can be
+eliminated by redeclaring the clashing name.*
+
+The effect of having an inline class declaration _DV_ with
+superinterfaces `V1, .. Vk` is that the members declared by _DV_ as
+well as all members of `V1, .. Vk` that are not redeclared by a
+declaration in _DV_ can be invoked on a receiver of the type
+introduced by _DV_.
+
+In the body of _DV_, a superinvocation syntax similar to an explicit
```

Comment:
I think I'm somewhat opposed to adding this. It feels pretty superfluous - you can always access a method on a super-interface `V` by doing `V v = this; v.method(...)`, or `(this as V).method(...)` or `V(r).method()` where `r` is the representation field.

Reply:
I like it.
We present the `inline class` as similar to a class, and it inherits implementation from super-views.

The obvious way to write a method in a sub-inline-class which does something more than the inherited method, is to use `before...;super.method(); after...`. If we can make it work for the common cases, I think our uses will appreciate it.

If there is a conflict on the member among the super-inline-classes (which can be the reason you wrote an overriding method to begin with, to resolve that conflict by forwarding to one of the super-classes), then you can't just use `super.method()`. That's sad, but unavoidable for now.
I don't want to introduce a way to choose a _specific_ super (at least not just for this, we can also use such a feature it in classes with multiple mixins), so in that case you will have to up-cast/inject into a super-interface before calling.

Reply:
I'm fine with allowing `super.foo`, but I'm reluctant to add the new `super.I.foo` syntax, so it sounds like we're on the same page here.

Reply:
It's the usual trade-off: If we support `super.I.foo` then we have a reasonably meaningful way to write forwarding methods that developers may need in order to resolve name clashes. If we do not support that syntax then we're encouraging developers to write code where type casts occur (and conflict resolution like this might not be so uncommon).

```dart
inline class V2 {
  final Object id;
  V2(this.id);
  void foo() { print('V2.foo()'); }
  void bar() { print('V2.bar()'); }
}

inline class V3 {
  final Object id;
  V3(this.id);
  void foo() { print('V3.foo()'); }
  void baz() { print('V3.baz()'); }
}

inline class V1 implements V2, V3 {
  final Object id;
  V1(this.id);
  void qux() {
    super.bar(); // 'V2.bar()'.
    super.baz(); // 'V3.baz()'.
    // super.foo(); // Compile-time error, do this instead:
    (this as V2).foo();
  }
}
```

Are you happy about telling developers to write `as`? Or do you have some other approach in mind?

Reply:
I'd like `I.super.foo` as a language feature. It makes sense for classes with multiple mixins.
Which means that when we add it, it should work there too, which means it's probably out-of-scope/feature-creep for the inline classes feature. 

(I'd recommend `V2(id).foo()` instead of using `as`, because `as` is unsafe - it doens't check that you're doing an upcast or valid cast at all. It also makes it very clear that the call will not have access to any members of `V1`.)

Reply:
Took the feature out. Resolving.